### ISSUE TITLE
Generics in function consts cause a parse error

### ISSUE THREAD

**Issue by lmanzke** [vuejs/vue-loader#1895]
### Version
16.8.2

### Reproduction link
[github.com](https://github.com/lmanzke/parser-repro)

### Steps to reproduce
1. Clone the repo
2. Install deps via yarn (it is a minimal project created with vue cli)
3. run lint (e.g. via npm run lint)

### What is expected?
No parse error and warning about incorrectly formatted parts

### What is actually happening?
When running the linter, the following error message appears: 

![Here is how it looks in PHPStorm](https://user-images.githubusercontent.com/7486693/139555381-ce1e7216-b14a-4a6e-9eda-713d93d92087.png)

```Parsing error: Unexpected token. Did you mean `{'>'}` or `&gt;`? at src/App.vue:10:24.```

The offending code is the following: 

```
const a = <T>(v: T): T => {
  return v;
}
```

which is perfectly valid Typescript code.

Note that this breaks linting in the whole file, meaning other problems are not detected (because the parsing process as a whole fails)

---
Note that the same code in main.ts does not trigger an error. Therefore I suspect it has to do with the loader in some way - it seems to incorrectly pass the code on to the linter after extracting it from the SFC (although this code still works, compiling is not affected)

It is of course possible to work around this by importing a function of this form from a .ts file. Nevertheless, it is sometimes annoying and makes people question if they used the correct syntax.

It is related to the generic parameter in front of the parens, as removing that and using string instead T for example yields no error.

**Comment by kkuegler** [comment]
I don't think this problem is related to the vue-loader project, but it has an easy fix.

As per https://github.com/vuejs/eslint-plugin-vue/issues/1343#issuecomment-737211636 you can change the `.eslintrc.js` to disable JSX parsing. This way the parser does not get confused about the type parameter which could also be an JSX tag.

```
  ...
  parserOptions: {
    ecmaVersion: 2020,
    ecmaFeatures: {
      jsx: false,
    },
  },
  ...
```

**Comment by lmanzke** [comment]
Hey, thanks for the info and the workaround - didn't know that. For me that is perfectly fine. It is still confusing that this happens in a standard vue-cli generated project. Isn't there a deeper problem here? I mean, is there a way to fix the JSX Parser to correctly understand that syntax - or is it just ambigious? (Meaning there is no way to distinguish in all cases what the programmer meant: generic or JSX tag)

**Comment by kkuegler** [comment]
The TypeScript project indeed recognized this syntax as ambiguous https://github.com/microsoft/TypeScript/issues/15713#issuecomment-300545248
In the same Github issue there is another (TS source) workaround: adding a comma after the type parameter `<T,>` - making it unambiguous https://github.com/microsoft/TypeScript/issues/15713#issuecomment-499474386

I also see that this happening in a stock vue-cli project is not optimal. However, JSX is enabled in the babel preset used by vue-cli by default. (see https://github.com/vuejs/vue-cli/tree/dev/packages/@vue/babel-preset-app#jsx)
This is a good argument for not adding the `jsx: false` option to the generated `.eslintrc.js` which could otherwise resolve the issue.

I hope this is happens rarely and people will find one of the Github issues pointing to this solution fast enough if they run into this problem :)
